std/experimental/allocator/building_blocks/affix_allocator.d

```diff
@@ -60,61 +63,103 @@ struct AffixAllocator(Allocator, Prefix, Suffix = void)
+                    {
+                        import core.atomic : cas;
+                        while (!cas(&_lockVar, true, false))
+                        {
+                            // Busy-wait for the parent to be set
+                        }
+                    }
+
+                    @nogc nothrow pure @trusted
+                    private static void parentUnlock(ref shared bool _lockVar)
+                    {
+                        import core.atomic : atomicStore;
+                        atomicStore(_lockVar, true);
+                    }
                 }
 
-                if (_parent.isNull)
+                @nogc nothrow pure @safe
+                Allocator parent()
                 {
-                    // If the `_parent` allocator is `null` we will assign
-                    // an object that references the GC as the `parent`.
-                    auto fn = (() @trusted =>
-                            cast(RCIAllocator function() @nogc nothrow pure @safe)(&wrapAllocatorObject))();
-                    _parent = fn();
+                    static @nogc nothrow
+                    RCIAllocator wrapAllocatorObject()
+                    {
+                        import std.experimental.allocator.gc_allocator : GCAllocator;
+                        import std.experimental.allocator : allocatorObject;
+
+                        return allocatorObject(GCAllocator.instance);
+                    }
+
+                    static @nogc nothrow
+                    RCISharedAllocator wrapProcAllocatorObject()
+                    {
+                        import std.experimental.allocator.gc_allocator : GCAllocator;
+                        import std.experimental.allocator : sharedAllocatorObject;
+
+                        return sharedAllocatorObject(GCAllocator.instance);
+                    }
+
+                    static if (is(Allocator == RCIAllocator))
+                    {
+                        if (_parent.isNull)
+                        {
+                            // If the `_parent` allocator is `null` we will assign
+                            // an object that references the GC as the `parent`.
+                            auto fn = (() @trusted =>
+                                    cast(RCIAllocator function() @nogc nothrow pure @safe)(&wrapAllocatorObject))();
+                            _parent = fn();
+                        }
+                    }
+                    else
+                    {
+                        parentLock(_lockVar);
+                        if (_parent.isNull)
+                        {
+                            auto fn = (() @trusted =>
+                                    cast(RCISharedAllocator function() @nogc nothrow pure @safe)
+                                    (&wrapProcAllocatorObject))();
+                            _parent = fn();
+                        }
+                        parentUnlock(_lockVar);
```

Comment:
To avoid deadlock instead put `scope(exit) parentUnlock(_lockVar);` right after the lock is acquired.